### ISSUE TITLE
Button text does not respect Space setting of DynamicFont tres

### ISSUE THREAD

**Issue by ErwinBr** [godotengine/godot#38630]
Godot version:
3.2.1 Stable

OS/device including version:
Mac OS Catalina 10.15.2

Issue description:
It seems that text on Buttons do not respect the Space setting of my DynamicFont. Screenshot shows my button (top) and label (bottom). The label is correct.

![Screenshot 2020-05-09 at 14 44 00](https://user-images.githubusercontent.com/46350573/81501438-ab22d780-92d8-11ea-988c-267ac6e1f55e.png)

The space is set to -20 on the TRES file. I'm 100% sure I'm using the same TRES for Button and Label text, because I only have one. Other properties, such as Size and Outline Size are working fine. 

![Screenshot 2020-05-09 at 16 39 25](https://user-images.githubusercontent.com/46350573/81501471-d9a0b280-92d8-11ea-94af-013497aa2e18.png)

Steps to reproduce:
1. Create a Button and a Label.
2. Create a DynamicFont and change the Spacing to -20
3. Assign the same DynamicFont file to Button and Label
4. Observe the difference

**Comment by follower** [comment]
Somewhat ironically, the reason `Label` displays the text correctly appears to be because `Label` *doesn't draw spaces*. :D

### Character width from `draw_char()` vs `get_char_size()`

Because `Label` doesn't draw spaces it never uses the width value returned for the space character from `draw_char()`, instead it uses the width value returned from `get_char_size()`.

And while `DynamicFont`'s `get_char_size()` implementation accounts for extra space character spacing its `draw_char()` implementation does *not* account for extra space character spacing.

### Other Affected code

This means that any code that uses `DynamicFont`'s `draw_char()` or `draw()` method to draw space characters or strings containing space characters will render the text incorrectly.

This also affects features like text alignment or clipping if the code uses width value(s) from `get_string_size()`/`get_char_size()` with width value(s) returned from `draw_char()` in calculations.

### Other handling of extra spacing

`DynamicFont`'s `draw_char()` implementation *does* account for extra *character* spacing, in addition to the width value returned from FreeType.

As a comparison, `BitmapFont`'s `draw_char()` implementation actually returns the result of `get_char_size()`.

### Solutions

This suggests that some potential solutions:

 * Modify `DynamicFont`'s `draw_char()` implementation to return the result of `get_char_size()` also although that might be considered "inefficient".

 * Modify `DynamicFont`'s `draw_char()` implementation to add to the returned width if the character is a space.

    But this duplicates the handling code and runs the risk that a similar situation might occur in future (or already occurs) if `get_char_size()` calculates anything differently to `draw_char()` (e.g. kerning? hinting? outlines?)

### Code to reproduce

Prerequisites:

 * A `DynamicFont` resource. (e.g. located at `res://fonts/barlow-regular-64.tres`)
 * The following script attached to a `Control` in a scene.

```gdscript
extends Control

var the_font: DynamicFont = preload("res://fonts/barlow-regular-64.tres")

func _ready():
    self.the_font.set_spacing(DynamicFont.SPACING_SPACE, -20)

func _draw():
    assert(the_font.get_string_size(" ").x == self.draw_char(the_font, Vector2(0,0), " ", ""))
```

Result:

 * The assertion in `_draw()` will fail.

### Red Herrings Encountered

From initial research, it appeared the issue may have been because `Label` & `Button` use two different ways of drawing their text:

 * `Label` uses `draw_char()` method of `FontDrawer`: https://github.com/godotengine/godot/blob/6a0473bcc23c096ef9ee929632a209761c2668f6/scene/gui/label.cpp#L278
 * `Button` uses `draw()` method of `Font` directly: https://github.com/godotengine/godot/blob/6a0473bcc23c096ef9ee929632a209761c2668f6/scene/gui/button.cpp#L228

But this difference was *not* the cause of the issue with extra space character spacing.

*[WIP: Intend to add some other info & code references but since "draft" issue comments aren't supported am posting this first.]*

**Comment by ErwinBr** [comment]
@akien-mga I'm not sure if this issue is supposed to be fixed in 3.2.4 (is it even the intention? I keep reading "doc clarity ..." not sure what it means?), but in any case I tested the RC1 and it does not seem to be different.

**Comment by ErwinBr** [comment]
I guess this issue will not be solved in 3.2.4?

**Comment by Calinou** [comment]
> I guess this issue will not be solved in 3.2.4?

Nobody has opened a pull request to fix the issue yet, so probably not. Also, since 3.2.4 is in RC stage, we can't risk merging anything that can cause regressions.

**Comment by ErwinBr** [comment]
> [The only pull request that was merged](https://github.com/godotengine/godot/pull/38683) is for documentation, not code.

Clear now, thanks. Maybe @bruvzg can have a look at this, since he is working on fonts for 4.0.

**Comment by Calinou** [comment]
This was likely fixed in 4.0 thanks to https://github.com/godotengine/godot/pull/46043.

**Comment by akien-mga** [comment]
Fixed by #46043 and #46055.
